CSS grid should not eagerly evaluate calc() for repetitions value

weinig opened the pull request:
#### 71e24b22e1d321b20aa881c422c45ed01674fc4b
<pre>
CSS grid should not eagerly evaluate calc() for repetitions value
<a href="https://bugs.webkit.org/show_bug.cgi?id=279308">https://bugs.webkit.org/show_bug.cgi?id=279308</a>

Reviewed by Darin Adler.

Replaces eager evaluation of calc() in CSS grid consumers with storing
the primitive value for resolution at style building time.

* LayoutTests/fast/css-grid-layout/grid-repeat-calc-expected.txt:
* LayoutTests/fast/css-grid-layout/grid-repeat-calc.html:
    - Adds tests for more calc() cases including use of relative-lengths and corrected
      existing test to ensure calc() is retained for the specified value.

* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
* Source/WebCore/css/CSSGridIntegerRepeatValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
    - Store repetitions as a CSSPrimitiveValue to all it to retain calc().

* Source/WebCore/inspector/InspectorOverlay.cpp:
    - The use of `resolveAsIntegerDeprecated` is quite unfortunate. This code should
      probably not be using pre-style resolved inline styles.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Resolve and clamp repetitions here now.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - The unit tests need to be able to call resolveAsIntegerNoConversionDataRequired(),
      a few inline/export changes are needed here.

* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
    - Use resolveAsIntegerNoConversionDataRequired() to resolve the now
      CSSPrimitiveValue repetitions value. This is ok since we know all
      the cases have no relative lengths.

Canonical link: <a href="https://commits.webkit.org/283350@main">https://commits.webkit.org/283350@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af5b6cc7a5c4ad99bdaee11c017f1fa0e83a49d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45426 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70084 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16662 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68171 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16944 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53022 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11603 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69120 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41915 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57183 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/33656 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38586 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15539 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60476 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14907 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71787 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10008 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/14320 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/60335 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10040 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57245 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60626 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14568 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8266 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1905 "Passed tests") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41234 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42310 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
```diff
@@ -171,7 +171,7 @@ static unsigned computeNumberOfTracks(const CSSValueContainingVector& valueList)
         if (is<CSSGridIntegerRepeatValue>(value)) {
             auto& repeatValue = downcast<CSSGridIntegerRepeatValue>(value);
-            numberOfTracks += repeatValue.repetitions() * computeNumberOfTracks(repeatValue);
+            numberOfTracks += repeatValue.repetitions().resolveAsIntegerNoConversionDataRequired() * computeNumberOfTracks(repeatValue);
```
What guarantees no conversion data is required?

weinig replied:
This is an API test, and all the test cases don't require conversion data. If one is added that does, this will assert.

darinadler commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -1276,9 +1276,7 @@ inline std::optional<GridTrackList> BuilderConverter::createGridTrackList(Builde
-            Vector<String> names;
-            for (auto& namedGridLineValue : lineNamesValue->names())
-                names.append(namedGridLineValue);
+            Vector<String> names { lineNamesValue->names() };
```
Can we do this without a local variable?

weinig replied:
Indeed it can.

weinig committed:
CSS grid should not eagerly evaluate calc() for repetitions value
https://bugs.webkit.org/show_bug.cgi?id=279308

Reviewed by Darin Adler.

Replaces eager evaluation of calc() in CSS grid consumers with storing
the primitive value for resolution at style building time.

* LayoutTests/fast/css-grid-layout/grid-repeat-calc-expected.txt:
* LayoutTests/fast/css-grid-layout/grid-repeat-calc.html:
    - Adds tests for more calc() cases including use of relative-lengths and corrected
      existing test to ensure calc() is retained for the specified value.

* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
* Source/WebCore/css/CSSGridIntegerRepeatValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
    - Store repetitions as a CSSPrimitiveValue to all it to retain calc().

* Source/WebCore/inspector/InspectorOverlay.cpp:
    - The use of `resolveAsIntegerDeprecated` is quite unfortunate. This code should
      probably not be using pre-style resolved inline styles.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Resolve and clamp repetitions here now.

* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
    - The unit tests need to be able to call resolveAsIntegerNoConversionDataRequired(),
      a few inline/export changes are needed here.

* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
    - Use resolveAsIntegerNoConversionDataRequired() to resolve the now
      CSSPrimitiveValue repetitions value. This is ok since we know all
      the cases have no relative lengths.

Canonical link: https://commits.webkit.org/283350@main
